### ISSUE TITLE
Pylance intellisense does not highlight object methods in case of its creation with different arg types

### ISSUE THREAD

**Issue by wvolkov** [microsoft/pylance-release#1869]
## Environment data

-   Language Server version: 2021.9.3 (pyright d2771b18)
-   OS and version: Ubuntu 20.04
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7

## Code Snippet / Additional information

Assume class `DecisionEngine`, which creates an instance of `DBClient`, passing an `integer` instead of `string` as specified below:

```python
class DBClient:

    def __init__(self, param: str):
        self.param = param

    async def setup(self):
        pass

class DecisionEngine:

    def __init__(self):
        self.db: DBClient = None

    async def setup(self):
        self.db = DBClient(1)
        await self.db.setup()
```

## Expected behaviour

at setup method of `DecisionEngine` class it is possible to drill down into `setup` method of `db` object:
![image](https://user-images.githubusercontent.com/49549154/134643410-29ee0664-e722-4c34-8210-3d7559a45f58.png)

## Actual behaviour

it interprets `db` object as `Any` type
![image](https://user-images.githubusercontent.com/49549154/134643501-2eefa573-ef8a-468c-9faf-b236773082a6.png)
despite the fact that it is correctly recognize the type of `db` object at upper line of code
![image](https://user-images.githubusercontent.com/49549154/134643592-1f5f69fa-e156-4e14-a8ea-5d49d4b83651.png)

**Comment by jakebailey** [comment]
This seems like a bug in some async handling, at least in the hover; repeat the same thing and I get:

![image](https://user-images.githubusercontent.com/5341706/134727039-2c023b6d-947c-42c3-a94e-54a58773a813.png)

Making `setup` return something lets me see:

![image](https://user-images.githubusercontent.com/5341706/134727159-6d27e0aa-b29f-43f7-8db2-2c23e6b2b715.png)

**Comment by erictraut** [comment]
First, there's a type error in the sample above. The expression `DBClient(1)` isn't valid because the constructor for `DBClient` accepts a `str`, not an `int`. If you enable type checking, this error will be flagged.

Second, I'm not able to repro the behavior you're seeing @jakebailey. I can't see all of your code, so perhaps there are differences? Where is the value of 1234 coming from? Here's what I see:

![Screen Shot 2021-09-24 at 1 38 20 PM](https://user-images.githubusercontent.com/7040122/134737055-6c274470-752e-42fe-b607-dc6fe58bc812.png)

```python
from typing import Optional

class DBClient:
    def __init__(self, param: str):
        self.param = param

    async def setup(self):
        pass

class DecisionEngine:
    def __init__(self):
        self.db: Optional[DBClient] = None

    async def setup(self):
        self.db = DBClient("1")
        reveal_type(await self.db.setup())
```

**Comment by jakebailey** [comment]
The code I have is this:

```
class DBClient:
    def __init__(self, param: str):
        self.param = param

    async def setup(self):
        return 1234

class DecisionEngine:
    def __init__(self):
        self.db: DBClient = None

    async def setup(self):
        self.db = DBClient(1)


        self.db.setup()
        await self.db.setup()

        reveal_type(self.db.setup())
        reveal_type(await self.db.setup())
```

Revealing the type produces the right result, but try hovering on `db` inside of the await and compare it to outside of the await. Hovering on it within `reveal_type` also says it's `Any`.

**Comment by jakebailey** [comment]
I did a bit of debugging, and our hover code sees that it's a name node, then asks for declarations from the type evaluator, but when it does getType, it returns undefined. (That's as far as I went to debug before seeing it was going to be an evaluator thing.)

**Comment by wvolkov** [comment]
> First, there's a type error in the sample above. The expression `DBClient(1)` isn't valid because the constructor for `DBClient` accepts a `str`, not an `int`. If you enable type checking, this error will be flagged.
> 
> Second, I'm not able to repro the behavior you're seeing @jakebailey. I can't see all of your code, so perhaps there are differences? Where is the value of 1234 coming from? Here's what I see:
> 
> ![Screen Shot 2021-09-24 at 1 38 20 PM](https://user-images.githubusercontent.com/7040122/134737055-6c274470-752e-42fe-b607-dc6fe58bc812.png)
> 
> ```python
> from typing import Optional
> 
> class DBClient:
>     def __init__(self, param: str):
>         self.param = param
> 
>     async def setup(self):
>         pass
> 
> class DecisionEngine:
>     def __init__(self):
>         self.db: Optional[DBClient] = None
> 
>     async def setup(self):
>         self.db = DBClient("1")
>         reveal_type(await self.db.setup())
> ```

I didn't see any type error at VS Code, why?

![image](https://user-images.githubusercontent.com/49549154/134737787-dc62f6a5-d8e6-45ad-9bb2-bbf132e85cf8.png)

**Comment by jakebailey** [comment]
@wvolkov Type checking isn't enabled by default in Pylance; you need to set `python.analysis.typeCheckingMode` or set it in a `pyrightconfig.json`.

**Comment by jakebailey** [comment]
Ah, so fixing the type error fixes this. That's interesting.

![image](https://user-images.githubusercontent.com/5341706/134738644-88b949d6-5306-435c-9167-768ca4e8a310.png)

![image](https://user-images.githubusercontent.com/5341706/134738658-3386ada5-f135-41f4-bae9-b645180f612b.png)

I'm not really sure why that would be the case, if reveal_type gets the right answer, and it works here:

![image](https://user-images.githubusercontent.com/5341706/134738700-f9a1dbca-2f9b-47cf-b889-8db8b4d63ea8.png)

**Comment by erictraut** [comment]
Yes, good catch. There is a bug here that causes the hover type for `self.db` to appears as `Unknown` even though it is eventually resolved to `DBClient`. The problem is that the logic that evaluates the type for the `await` statement was not honoring the "incomplete" flag and was marking the type evaluation for itself as completely resolved. A fix for this bug will be included in the next release.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.9.4, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202194-29-september-2021
